utils/zerovec/src/varzerovec/owned.rs

```diff
@@ -0,0 +1,325 @@
+            entire_slice: Vec::new(),
+        }
+    }
+
+    pub fn from_components(components: SliceComponents<T>) -> Self {
+        Self {
+            marker: PhantomData,
+            entire_slice: components.entire_slice().into(),
+        }
+    }
+
+    /// Construct a VarZeroVecOwned from a list of elements
+    pub fn from_elements(elements: &[T]) -> Self {
+        Self {
+            marker: PhantomData,
+            entire_slice: components::get_serializable_bytes(elements).unwrap(),
+        }
+    }
+
+    /// Try to allocate a buffer with enough capacity for `capacity`
+    /// elements. Since `T` can take up an arbitrary size this will
+    /// just allocate enough space for 4-byte Ts
+    pub(crate) fn with_capacity(capacity: usize) -> Self {
+        Self {
+            marker: PhantomData,
+            entire_slice: Vec::with_capacity(capacity * 8),
+        }
+    }
+
+    /// Try to reserve space for `capacity`
+    /// elements. Since `T` can take up an arbitrary size this will
+    /// just allocate enough space for 4-byte Ts
+    pub(crate) fn reserve(&mut self, capacity: usize) {
+        self.entire_slice.reserve(capacity * 8)
+    }
+
+    pub(crate) fn get_components<'a>(&'a self) -> SliceComponents<'a, T> {
+        unsafe {
+            // safety: VarZeroVecOwned is guaranteed to parse here
+            SliceComponents::from_bytes_unchecked(&self.entire_slice)
+        }
+    }
+
+    /// Get the number of elements in this vector
+    pub fn len(&self) -> usize {
+        self.get_components().len()
+    }
+
+    /// Returns `true` if the vector contains no elements.
+    pub fn is_empty(&self) -> bool {
+        self.get_components().is_empty()
+    }
+
+    /// Obtain an iterator over VarZeroVecOwned's elements
+    pub fn iter<'b>(&'b self) -> impl Iterator<Item = &'b T::VarULE> {
+        self.get_components().iter()
+    }
+
+    /// Get one of VarZeroVecOwned's elements, returning None if the index is out of bounds
+    pub fn get(&self, idx: usize) -> Option<&T::VarULE> {
+        self.get_components().get(idx)
+    }
+
+    /// Get this [`VarZeroVecOwned`] as a borrowed [`VarZeroVec`]
+    ///
+    /// If you wish to repeatedly call methods on this [`VarZeroVecOwned`],
+    /// it is more efficient to perform this conversion first
+    pub fn as_varzerovec<'a>(&'a self) -> VarZeroVec<'a, T> {
+        VarZeroVec(VarZeroVecInner::Borrowed(self.get_components()))
+    }
+
+    /// Empty the vector
+    pub fn clear(&mut self) {
+        self.entire_slice.clear()
+    }
+
+    pub fn to_vec(&self) -> Vec<T>
+    where
+        T: Clone,
+    {
+        self.get_components().to_vec()
+    }
+
+    /// Insert an element at index `idx`
+    pub fn insert(&mut self, index: usize, element: &T) {
+        let len = self.len();
+        if index > len {
+            panic!(
+                "Called out-of-bounds insert() on VarZeroVec, index {} len {}",
+                index, len
+            );
```

Comment:
I don't really agree; I'd like to match `std::vec` on behavior here. We can also provide a `try_insert` but I'd rather track fallible APIs separately? Because adding them would involve ZeroMap changes too

Reply:
https://github.com/unicode-org/icu4x/issues/1081